Optsbuilder from hashmap

hpca01 opened the pull request:
Hi @blackbeam 

Added ability to set up from HashMap<String,String>, also added a test case in the opts.rs file to test the change.

Let me know if I missed something, my Rust is a bit rusty :laughing: 

Thanks for allowing me to contribute!

blackbeam commented on src/conn/opts.rs:
```diff
@@ -461,6 +461,43 @@ impl OptsBuilder {
+    /// Login .cnf file parsing lib https://github.com/rjcortese/myloginrs returns a HashMap for client configs
+    ///
+    /// **Note:** You do **not** have to use myloginrs lib.    
+    pub fn from_hash_map(mut self, client: &HashMap<String, String>) -> Self {
```
I think it'd be reasonable to either:
* rename it to `from_mylogin_cnf` because the list of supported options is limited by the `mysql_config_editor` capabilities;
* or to add support for [other url parameters](https://docs.rs/mysql/20.0.1/mysql/#url-based-connection-string).

I'd personally suggest the second option.

hpca01 replied:
I'll work to add the second option

blackbeam commented on src/conn/opts.rs:
```diff
@@ -461,6 +461,43 @@ impl OptsBuilder {
+                }
+                "socket" => self.opts.0.socket = Some(value.to_string()),
+                _ => {
+                    //do nothing
```
If it won't be renamed to `from_mylogin_cnf`, then could you please emit `UnknownParameter` here (and thus, make this function fallible)?

hpca01 replied:
I'll do the second option you suggested, to add in the other url params, it's better that way since that also makes it more useful to use that function outside of the mylogincnf lib.

Thanks for the suggestion!

hpca01 commented:
Hi @blackbeam 

I added the following:
- The other remaining URL params
- Throw `UnknownParameter` exception when a key passed in via HashMap is not recognized
- Adjusted the documentation for the function to include the additional URL params
- Adjusted the test case to test for the additional URL params.

Thanks! Let me know if there's any adjustments needed.

blackbeam commented on src/conn/opts.rs:
```diff
@@ -461,6 +461,84 @@ impl OptsBuilder {
+                    self.opts.0.ip_or_hostname = host;
+                }
+                "port" => {
+                    let parsed: u16 = value.parse::<u16>().unwrap_or(3306);
```
Could you please emit the `InvalidValue` for parse errors here and for other parameters? I'm trying to be as rigorous as possible here to avoid data loss, that is possible if errors like this are silently ignored.
The rule of thumb here is that if an option value is explicitly specified, then the client will either start with the specified value of an option or won't start at all

hpca01 replied:
@blackbeam 

Thanks for the feedback, I added `InvalidValue` to all parsing errors as suggested.
Also added:
- Another test case to assert for errors raised
- Added `PartialEq` to `Optsbuilder`, let me know if that is okay. If not, I can do a string check in the error case vs a type of error check.

Thanks again for feedback! It's my first time contributing to a project on git

blackbeam commented on src/conn/opts.rs:
```diff
@@ -461,6 +461,102 @@ impl OptsBuilder {
         OptsBuilder { opts: opts.into() }
     }
 
+    /// Use a HashMap for creating creating an OptsBuilder instance:
```
```suggestion
    /// Use a HashMap for creating an OptsBuilder instance:
```

blackbeam commented:
@hpca01, thanks!
